### PR TITLE
major refactoring for lazy/cached forward computation

### PR DESCRIPTION
**Pull request by autonull**
hi this is an experimental fork I started which I thought @pingng might be interested in.  it's not purely an improvement, as I've disabled most of the OpenCL support in the process.  rather it's more of an exploration.

the main change I've made so far is a lazy and cached __forward__ computation pass as seen in this example
https://github.com/automenta/tadlib/blob/main/src/test/java/com/codeberry/tadlib/tensor/TensorTrainSimple.java#L37

note the uppercase operators:
```
        Tensor y = ADD(MATMUL(x_data, w), b);
        Tensor diff = SUB(y, y_data);
        Tensor diffSq = MUL(diff, diff);
        Tensor diffSqSum = SUM(diffSq);
```

These invoke an alternate tensor constructor that uses the new lazy / cached forward pass.  Here is ADD as an example, from Ops.java:
```
   public static Tensor ADD(Tensor a, Tensor b) {
        return new Tensor(v->{
            v.set(a.val().add(b.val()));
        }, a.shape(), asList(parent(a, funcGradientAdd(a)), parent(b, funcGradientAdd(b))));
    }
```
note that the forward pass is now implemented as a lambda.  Internally the tensor stores a reverse link for invalidating the results of its dependents, recursively.

There's a bit more work I want to do towards compiling the entire forward and backward pass, and either reimplemnting the OpenCL or seeing if Aparapi can handle that for TADlib automatically.

Interested in your feedback @pingng .  Thanks for this great project.

### CODE REVIEW

**Comment by pingng** [comment]
Hi, your changes look interesting. It will be useful for people building large computation graphs but reading results requiring a portion of it.

The main code edits look good. I guess you are still experimenting, so it's natural that there are remnants of the opengl support code (like usage of disposeAllExceptReturnedValue()).

There are a couple of things that crossed my mind (though you might have already thought about this):

- What about selecting between 'plain computation' and 'record graph' mode? It 
can be a simple call:
 ```
 (JavaProvider or some other singleton).setComputationMode(PLAIN);
 ```
...or maybe using (nestable) try-with-resource syntax:
```
try((JavaProvider or some other singleton).useComputationMode(GRAPH)) {
    try((JavaProvider or some other singleton).useComputationMode(PLAIN)){
      fixedResult = add(a, b);
    }
    liveResult = add(fixedResult, b);
}
```
   Maybe there is no need for the plain old mode. I'm just thinking out loud :)
- Use weak references for holding the forward nodes to prevent memory leaks.

Aparapi looks like a handy framework. It would be fun if you got it working with the code. Thanks for your interest in the project. I hope you have fun with it :)

**Comment by autonull** [comment]
I like these suggestions, and I can tell you see similar potentials as I'm seeing.

Multiple backends are great, but I imagine them being selected by an adaptive AUTO mode, like a JIT with performance monitoring, measuring the given tensor graph and resource availability conditions.

At the same time, something like Aparapi or TornadoVM may be able to provide some or all of the CL backend support.

The latest changes I've commited are integration with the Optimizer interface, which affects "optimizable"-flagged tensors from the graph.  These are also targets for a weight initialization interface.

Ultimately I'd like this fork of TADLib to replace the inflexible code I've got here:
https://github.com/dumbai/jcog/tree/main/learn/src/main/java/jcog/nn
https://github.com/dumbai/jcog/blob/main/learn/src/test/java/jcog/learn/MLPTest.java

as I want to be able to translate pytorch code that makes use of complicated loss functions and arbitrary tensor graph shapes.  for example, reinforcement learning algorithms like PPO and A3C.

i think something of the scope of DL4J is possible with TADLib, but much simpler and smaller, and that fully supports CPU-only backends, which is actually the preferred target for some "low-latency" applications I need it for.

my next step is to start combining my existing MLP code and code which uses it, which for example includes an Adam optimizer that I'd port over.  For example, my MLP class implements a generic gradient-based Predictor interface, which my Q-Learning agent uses as its model.
https://github.com/dumbai/jcog/tree/main/learn/src/main/java/jcog/rl/dqn

There are some other projects you might be interested here:
https://gitlab.com/sseehh/metanarchy#install

**Comment by autonull** [comment]
Here is a test which involves all the major changes:
https://github.com/automenta/tadlib/blob/main/src/test/java/com/codeberry/tadlib/tensor/TensorTrainSimple.java#L74

**Comment by pingng** [comment]
I'm closing this pull request. Thanks for showing me your project. I hope TADLib is not too messy and is compatible with what you are trying to achieve. I'll try to check out your progress from time to time. Cheers!
